File: kotlin/src/main/kotlin/day01/SonarSweep2.kt
package day01

import Common

fun main() {
    val values = Common.readInput("/day01/input.txt").map(String::toInt)
    println(SonarSweep2().countIncreases(values))
}

class SonarSweep2 {

    fun countIncreases(values: List<Int>): Int {
        val values2 = values.subList(1, values.size)
        val values3 = values.subList(2, values.size)

        var increases = 0

        for (i in 1 until values3.size) {
            val window1 = values[i - 1] + values2[i - 1] + values3[i - 1]
            val window2 = values[i] + values2[i] + values3[i]

            if (window2 > window1) {
                increases++
            }
        }

        return increases
    }
}

File: scala/src/main/scala/day01/SonarSweep2.scala
package day01

import common.Common

object SonarSweep2 {

  def main(args: Array[String]): Unit = {
    val values = Common.readFile("/day01/input.txt").map { value => value.toInt }

    val values2 = values.drop(1)
    val values3 = values.drop(2)
    val values4 = values.drop(3)
    val slidingWindows1 = (values, values2, values3).zipped
    val slidingWindows2 = (values2, values3, values4).zipped
    val increases = slidingWindows1.zip(slidingWindows2)
      .count { case ((a, b, c), (u, v, x)) => a + b + c < u + v + x }

    println(increases)
  }
}
